lib/bolt_server/pe/pe_pal.rb

```diff
@@ -0,0 +1,135 @@
+        # we finish with the settings we need for PEBolt::PAL init.
+        with_puppet_settings { |_| nil }
+      end
+    end
+
+    def initialize(plan_executor_config, environment_name, hiera_config = nil, max_compiles = nil)
+      # Bolt::PAL#initialize takes the modulepath as its first argument, but we
+      # want to customize it later, so we pass an empty value:
+      super([], hiera_config, max_compiles)
+
+      codedir = plan_executor_config['codedir'] || DEFAULT_BOLT_CODEDIR
+      environmentpath = plan_executor_config['environmentpath'] || "#{codedir}/environments"
+      basemodulepath = plan_executor_config['basemodulepath'] || "#{codedir}/modules:/opt/puppetlabs/puppet/modules"
+
+      with_pe_pal_init_settings(codedir, environmentpath, basemodulepath) do
+        environment = Puppet.lookup(:environments).get!(environment_name)
+
+        # In the instance where the environment is "production" but no production dir
+        # exists, the lookup will succeed, but the configuration will be mostly empty.
+        # For other environments the lookup will fail, but for production we don't
+        # want cryptic messages sent to the user about combining `nil` with a string.
+        # Thus if we do get here and `path_to_env` is empty, just assume it's the
+        # default production environment and continue.
+        #
+        # This should hopefully match puppet's behavior for the default 'production'
+        # environment: _technically_ that environment always exists, but if the dir
+        # isn't there it won't find the module and fail with "plan not found" rather
+        # than "environment doesn't exist"
+        bolt_config = File.join(environment.configuration.path_to_env, 'bolt.yaml') if environment.configuration.path_to_env
+        # If we find a bolt.yaml config in the environment root, we load that instead of
+        # environment.conf to find the modulepath. modulepath will be the _only_ setting
+        # that will work from bolt.yaml in plans in PE.
+        if bolt_config && File.exists?(bolt_config)
+          modulepath_dirs = Bolt::Util.read_config_file(bolt_config)['modulepath'].split(File::PATH_SEPARATOR)
+        else
+          modulepath_dirs = environment.modulepath
+        end
+
+        # A new modulepath is created from scratch (rather than using super's @modulepath)
+        # so that we can have full control over all the entries in modulepath. In the future
+        # it's likely we will need to preceed _both_ Bolt::PAL::BOLTLIB_PATH _and_
+        # Bolt::PAL::MODULES_PATH which would be more complex if we tried to use @modulepath since
+        # we need to interject a path between *modulepath_dirs and Bolt::PAL::MODULES_PATH
+        @original_modulepath = modulepath_dirs
+        @modulepath = [PE_BOLTLIB_PATH, Bolt::PAL::BOLTLIB_PATH, *modulepath_dirs, Bolt::PAL::MODULES_PATH]
+      end
+    end
+
+    # There's a chance that some plan functions could conflict with built-in
```

Comment:
Do we actually need this in order to load metadata?

Reply:
Ah, yeah we can definitely delete the next part

Reply:
Is the value here on keeping pe_pal.rb identical with orchestrator's, or only putting the necessary parts in here?

Reply:
IMO just the necessary parts.